### ISSUE TITLE
Is AWX truly namespaced vs cluster scoped?

### ISSUE THREAD

**Issue by ddevalco** [ansible/tower-operator#705]
I understand that Operator needs to be installed and uses cluster scopes for:

customresourcedefinitions, clusterroles, and clusterrole bindings as those inherently are all cluster-scoped

However this PR:  https://github.com/ansible/awx-operator/pull/541 talks about moving from Clusterscoped to namespaced-scoped.

When I do an Operator install using 0.15.0, I am seeing failures related to all the things that are cluster scoped or do I have a misunderstanding of what that means?

You seem to need more than just Project permissions but rather Cluster admin permissions to apply the following:

```
customresourcedefinition.apiextensions.k8s.io/awxbackups.awx.ansible.com
customresourcedefinition.apiextensions.k8s.io/awxrestores.awx.ansible.com
customresourcedefinition.apiextensions.k8s.io/awxs.awx.ansible.com
clusterrole.rbac.authorization.k8s.io/awx-operator-metrics-reader
clusterrole.rbac.authorization.k8s.io/awx-operator-proxy-role
clusterrolebinding.rbac.authorization.k8s.io/awx-operator-proxy-rolebinding
```

Errors from Operator install:

```
make deploy
cd config/manager && /Users/ddevalco/Dropbox/escm/awx19_5/awx-operator/bin/kustomize edit set image controller=quay.io/ansible/awx-operator:0.15.0
cd config/default && /Users/ddevalco/Dropbox/escm/awx19_5/awx-operator/bin/kustomize edit set namespace escm-stg
/Users/ddevalco/Dropbox/escm/awx19_5/awx-operator/bin/kustomize build config/default | kubectl apply -f -
namespace/escm-stg unchanged
serviceaccount/awx-operator-controller-manager unchanged
configmap/awx-operator-awx-manager-config created
service/awx-operator-controller-manager-metrics-service created
deployment.apps/awx-operator-controller-manager created
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
Name: "awxbackups.awx.ansible.com", Namespace: ""
from server for: "STDIN": customresourcedefinitions.apiextensions.k8s.io "awxbackups.awx.ansible.com" is forbidden: User "u-w4453qrfpa" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
Name: "awxrestores.awx.ansible.com", Namespace: ""
from server for: "STDIN": customresourcedefinitions.apiextensions.k8s.io "awxrestores.awx.ansible.com" is forbidden: User "u-w4453qrfpa" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
Name: "awxs.awx.ansible.com", Namespace: ""
from server for: "STDIN": customresourcedefinitions.apiextensions.k8s.io "awxs.awx.ansible.com" is forbidden: User "u-w4453qrfpa" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope
Error from server (Forbidden): error when creating "STDIN": roles.rbac.authorization.k8s.io "awx-operator-awx-manager-role" is forbidden: user "u-w4453qrfpa" (groups=["azuread_group://"system:authenticated" "system:cattle:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:[""], Resources:["events"], Verbs:["delete"]}
{APIGroups:[""], Resources:["pods/log"], Verbs:["create"]}
{APIGroups:[""], Resources:["rolebindings"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:[""], Resources:["roles"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:[""], Resources:["services/finalizers"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["apps"], Resources:["deployments/finalizers"], ResourceNames:["awx-operator"], Verbs:["update"]}
{APIGroups:["apps"], Resources:["ingresses"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["awx.ansible.com"], Resources:["*"], Verbs:["*"]}
{APIGroups:["awx.ansible.com"], Resources:["awxbackups"], Verbs:["*"]}
{APIGroups:["awx.ansible.com"], Resources:["awxrestores"], Verbs:["*"]}
{APIGroups:["networking.k8s.io"], Resources:["daemonsets"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["networking.k8s.io"], Resources:["deployments"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["networking.k8s.io"], Resources:["replicasets"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["networking.k8s.io"], Resources:["statefulsets"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["configmaps"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["endpoints"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["events"], Verbs:["get" "list" "delete" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["persistentvolumeclaims"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["pods"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["secrets"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["serviceaccounts"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["services"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["services/finalizers"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["route.openshift.io"], Resources:["routes"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
{APIGroups:["route.openshift.io"], Resources:["routes/custom-host"], Verbs:["get" "list" "create" "delete" "patch" "update" "watch"]}
Error from server (Forbidden): error when creating "STDIN": roles.rbac.authorization.k8s.io "awx-operator-leader-election-role" is forbidden: user "u-w4453qrfpa" (groups=["azuread_group:// "system:authenticated" "system:cattle:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:["coordination.k8s.io"], Resources:["leases"], Verbs:["get" "list" "watch" "create" "update" "patch" "delete"]}
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "rbac.authorization.k8s.io/v1, Resource=clusterroles", GroupVersionKind: "rbac.authorization.k8s.io/v1, Kind=ClusterRole"
Name: "awx-operator-metrics-reader", Namespace: ""
from server for: "STDIN": clusterroles.rbac.authorization.k8s.io "awx-operator-metrics-reader" is forbidden: User "u-w4453qrfpa" cannot get resource "clusterroles" in API group "rbac.authorization.k8s.io" at the cluster scope
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "rbac.authorization.k8s.io/v1, Resource=clusterroles", GroupVersionKind: "rbac.authorization.k8s.io/v1, Kind=ClusterRole"
Name: "awx-operator-proxy-role", Namespace: ""
from server for: "STDIN": clusterroles.rbac.authorization.k8s.io "awx-operator-proxy-role" is forbidden: User "u-w4453qrfpa" cannot get resource "clusterroles" in API group "rbac.authorization.k8s.io" at the cluster scope
Error from server (NotFound): error when creating "STDIN": roles.rbac.authorization.k8s.io "awx-operator-awx-manager-role" not found
Error from server (NotFound): error when creating "STDIN": roles.rbac.authorization.k8s.io "awx-operator-leader-election-role" not found
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "rbac.authorization.k8s.io/v1, Resource=clusterrolebindings", GroupVersionKind: "rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding"
Name: "awx-operator-proxy-rolebinding", Namespace: ""
from server for: "STDIN": clusterrolebindings.rbac.authorization.k8s.io "awx-operator-proxy-rolebinding" is forbidden: User "u-w4453qrfpa" cannot get resource "clusterrolebindings" in API group "rbac.authorization.k8s.io" at the cluster scope
make: *** [deploy] Error 1

```

I guess I'm confused about this PR I mentioned above and the fact it alludes to moving away from anything cluster scoped, yet I'm still seeing errors that indicate you can't really completely move everything from being cluster scoped?

Just can't seem to get past these errors to install via Operator and wondering what am I missing understanding about this.

**Comment by jCOTINEAU** [comment]
i am not a contributor, but from my understanding and what we are achieving i can explain this.

Because CRDs definitions are cluster wide, you would always needs some more role before the oprator can be deployed ( or have the crds created by an admin ).

The thing that is really namespace scoped now is that the operator itself will not be watching crds instance ( awxs, awxbackup, awxrestore ) cluster widely. but will list them in his own namespace. ( before it was watching at some kind of cluster level )

But you are right there is also some clusterRole, that are not namespace-scoped ( rbac-proxy and reader ), which from my point of view are not very useful or should be configurable ( rbac-proxy is useful only when you have no netpols )

I find this a little bit sad as it prevent from having one operator manage multiple awx instanced in multiple namespaces.

It would have been better to have a ClusterRole ( created by admin ) rolebinded to each awx namespaces, with a way to let the opeator be aware of those namespaces.

The rbac system of kubernetes is still not perfect, and everyone is building around the best they can.

In conclusion, operator cannot be fully deployed by a user with only rights on a namespace.

PS: remember the operator is in alpha, it is already quite cool and will be improved later on.
PS2: sorry for my bad english hope it was kind of useful informations

**Comment by hungpr0** [comment]
From what I see, PR https://github.com/ansible/awx-operator/pull/541 isn't really namespaced scope. You need cluster admin right to create CRD in any way. What the PR does is just limiting the namespace that can only deploy `awx-operator`, other namespaces can't deploy the operator. 

```
Starting with awx-operator 0.14.0, AWX can only be deployed in the namespace that the operator exists in. This is called a namespace-scoped operator
```

**Comment by kincerb** [comment]
I don't know if this is a similar problem to what I'm experiencing, but I cannot deploy because of the CRDs. My company has a managed kubernetes (rancher), and I do not have the permissions to create custom resources since they are cluster scoped.

I would really be interested if there are any known workarounds. I don't quite understand the assumption of cluster admin rights. That being said, I am very new to kubernetes.

**Comment by TheKangaroo** [comment]
@kincerb this is basically the same issue, yes. There are two types of resources in kubernetes, namespaced and cluster scoped. CustomResourceDefinitions are cluster scoped and therefore, can only be created with the cluster admin permission (or at least read/write CRD permission).
There is not really a workaround for this issue. You need more permissions if you want to install CRDs (and cluster scoped RBAC Resources). I'm from the cluster management team in our company and we used to have different strategies to deal with this:
* we did not allow any operators in a multi tenant cluster at all (not what you want ;) )
* we (the cluster admin team) installed operators for cluster users (this needs a lot of syncing/work from both sides, but this is the only way to make it work without breaking the authorization concept). We carefully checked every single operator and try to use as little operators as possible.
* we provide clusters for powerusers for this exact use case, where users get cluster admin permission on their cluster and install operators by themself.
* (your cluster admins _cloud_ give you read/write access to CRDs only, but as mentioned before you would also need to create cluster scoped RBAC resources and creating RBAC would allow you to grant cluster admin to yourself and nothing has been gained, so we never considered this)

I cannot think of any other solution. I think this isn't  exactly what you hoped for but I think this is as close as you can get.

**Comment by kincerb** [comment]
@TheKangaroo , thank you very much for clarifying this for me. I hope this helped others as well. I appreciate you taking the time to educate me to understand this.

**Comment by br0nwe** [comment]
Plus one here.  inconsistent scoping messes with our tenant based cluster restriction
